### PR TITLE
Medmi pollen and refactor - Re-add extra datasets and removed timeseries

### PR DESCRIPTION
**Pull request by AnnAnnFryingPan**
Re-add extra dataset functionality - removed timeseries attempts, which is now in branch dev_timeseries

### CODE REVIEW

File: MEDMI_Data_Download/pollen_site_met_calculations.py [hunk 1]
```diff
@@ -22,8 +22,8 @@
 pollen_sites['WORCESTER']       = (52.19716, -2.24165, 40.0)
 pollen_sites['YORK']            = (53.94819,   -1.05194, 25.0)
 
-Dates = ['2016-1-1 0','2019-12-31 23']
-Dates_string = '2016-2019'
+Dates = ['2017-1-1 0','2017-1-1 23']
+Dates_string = '2017-2017'
```
Comment (douglowe) [inline MEDMI_Data_Download/pollen_site_met_calculations.py]: should we not keep the 2016 - 2019 dates?

File: MEDMI_Data_Download/met_extraction_script.py [hunk 2]
```diff
@@ -227,8 +492,7 @@ def create_directory(dir_name):
         if len(args.date_range) >= 2:
```
Comment (douglowe) [inline MEDMI_Data_Download/met_extraction_script.py:514]: should we have a warning here - telling users that only the first 2 dates are being used?

File: MEDMI_Data_Download/met_extraction_script.py [hunk 3]
```diff
@@ -166,19 +411,25 @@ def create_directory(dir_name):
     parser = argparse.ArgumentParser(
         description="*** A script for automated downloading of MET data for a given date range. ***")
 
-    AVAILABLE_MEASUREMENTS = ['pollen', 'rain', 'temp',  'wind']
+    AVAILABLE_MEASUREMENTS = ['rain', 'temp', 'rel_hum', 'pressure', 'dewpoint', 'wind', 'pollen']
+    AVAILABLE_EXTRA_MEASUREMENTS = \
+        ['rain', 'temp', 'rel_hum', 'pressure', 'dewpoint', 'wind']# + MetExtractorPollen.pollen_species
+                                                                    # Pollen species input as extra data fail silently
+                                                                    # (the output just has no column for that data)
```
Comment (douglowe) [inline MEDMI_Data_Download/met_extraction_script.py]: Yep - because the stations which measure pollen don't measure meteorological parameters (and vice versa).

We should add some notes in the README regarding the use of extra measurements - telling the user how this is calculated (and why they should pick the most common measurement as the main variable if possible, so that they lose the least data).

File: MEDMI_Data_Download/met_extraction_script.py [hunk 4]
```diff
@@ -48,107 +163,237 @@ def _extract_data(self, dict, settings, extra_datasets=[]):
...
+
+class MetExtractorDewpoint(MetExtractor):
+    SOURCE_REFERENCE = 'midas.weather_hrly_ob.dewpoint'
+    MEASUREMENT_NAME = 'dewpoint'
+
+    def __init__(self, out_dir=MetExtractor.DEFAULT_OUT_DIR, verbose=MetExtractor.DEFAULT_VERBOSE):
+        super(MetExtractorDewpoint, self).__init__(out_dir, verbose)
+        self._head_string = 'Dewpoint hourly data {} for date range: {} to {}\n'
+        self._cols_specific = [MetExtractorDewpoint.MEASUREMENT_NAME]
+        self._filename = '{}/wet_bulb{}{}.csv'.format(self._out_dir, '{}', '{}')
```
Comment (douglowe) [inline MEDMI_Data_Download/met_extraction_script.py]: 'wetbulb' here is a hangover from when I was pulling out the wetbulb temperature (which is different to the dewpoint temperature). The wetbulb temperature can be used to calculate the relative humidity, but the equation required takes both temperature and pressure as inputs too (if I recall correctly) - and in the dataset we have some relative humidity data where there is no pressure reading (so I don't think this was the equation used by the met office). So I switched to extracting the dewpoint temperature instead (which can be used to calculate RH with only temperature, but no pressure, data), but forgot to change the name of the output file too :/

tl:dr - could you change the output file name to 'dewpoint' instead?

File: MEDMI_Data_Download/pollen_site_met_calculations.py (hunk 1 above)
Comment (douglowe) [reply MEDMI_Data_Download/pollen_site_met_calculations.py]: Though - if this has been folded into the main script, should we just delete this script?

File: MEDMI_Data_Download/met_extraction_script.py [hunk 5]
```diff
@@ -166,19 +411,25 @@ def create_directory(dir_name):
...
-    # Measurement
+    # Measurements
     parser.add_argument("--measurements", "-m", metavar='M', type=str, nargs='+', help="measurements to be extracted. \
             Must be in (and defaults to) {}".format('[' + ", ".join([measure for measure in AVAILABLE_MEASUREMENTS]) + ']'))
-    #parser.add_argument("--measurements", "-m", dest="measurement", type=str, nargs='+',
-    #                    help="measurements to be extracted. Default: all")
+
+    parser.add_argument("--extra_measurements", "-x", metavar='X', type=str, nargs='+', help="extra measurements to be \
+        extracted along with each of the main measurements. Must be in {}. Default: None".format(
+        '[' + ", ".join([measure for measure in AVAILABLE_EXTRA_MEASUREMENTS]) + ']'))
```
Comment (douglowe) [inline MEDMI_Data_Download/met_extraction_script.py]: It doesn't make sense to me to apply the extra measurements to all the requested measurements. Can we change this so that you specify which measurement you want to add the extra measurements to? e.g.
--measurements temp rain pollen wind
--extra_measurements temp pressure dewpoint rel_hum
*Or*, we add an extra flag to call to indicate which measurements need the addition:
--measurements temp rain pollen wind
--add_to_measurement temp
--extra_measurements pressure dewpoint rel_hum

File: MEDMI_Data_Download/met_extraction_script.py [hunk 6]
```diff
@@ -48,107 +163,237 @@ def _extract_data(self, dict, settings, extra_datasets=[]):
...
-        extra_datasets = ['midas.weather_hrly_ob.rltv_hum', 'midas.weather_hrly_ob.stn_pres',
-                          'midas.weather_hrly_ob.dewpoint']
-        print('extracting {}'.format(self.head_string.format(date_range[0], date_range[1])))
-        dict = self._create_dict_base(date_range, latitude_range, longitude_range)
-        dict.update({'Source reference': self.source_reference})
-        settings = {'fname': filename,
-                    'headstring': self.head_string.format(date_range[0], date_range[1]),
-                    'columnstring': column_string}
-        self._extract_data(dict, settings, extra_datasets)
+    DEFAULT_PROCESS_SPATIAL_TEMPERATURE = 'sp_altadj_idw_mean'
```
Comment (douglowe) [inline MEDMI_Data_Download/met_extraction_script.py]: I don't think we should include the spatial processing (as this was for the pollen timeseries, which we aren't going to deal with anymore)?

File: MEDMI_Data_Download/met_extraction_script.py [hunk 7]
```diff
@@ -48,107 +163,237 @@ def _extract_data(self, dict, settings, extra_datasets=[]):
...
+        'ESKDALEMUIR': (55.31184, -3.20545, 236.0),
+        'EXETER': (50.73599, -3.53101, 10.0),
+        'IPSWICH': (52.05638, 1.1997, 41.0),
+        'LEICESTER': (52.62155, -1.12227, 91.0),
+        'LONDON': (51.51022, -0.11533, 45.0),
+        'MYLNEFIELD': (56.45699, -3.07182, 31.0),
+        'PLYMOUTH': (50.37461, -4.13725, 45.0),
+        'WIGHT': (50.71052, -1.29944, 32.0),
+        'WORCESTER': (52.19716, -2.24165, 40.0),
+        'YORK': (53.94819, -1.05194, 25.0)}
```
Comment (douglowe) [inline MEDMI_Data_Download/met_extraction_script.py]: take out pollen_sites, as we don't use this anymore

File: MEDMI_Data_Download/met_extraction_script.py (hunk 5 above)
Comment (AnnAnnFryingPan) [reply MEDMI_Data_Download/met_extraction_script.py]: To make it easier (for us at this point), we could just have an '--add_extra_datasets' flag that adds all extra datasets that are realistic for the requested measurement(s) and explain these in Readme. So
i) there is only one parameter, so less work required on parsing and matching measurements with extras
ii) measurements like pollens would not have any extras datasets
iii) Temp for example would have: humidity, dewpoint and pressure

Comment (douglowe) [reply MEDMI_Data_Download/met_extraction_script.py]: Okay - but can we restrict the extra datasets to just temp, pressure, dewpoint, and rel_hum? Both to as the extra datasets which are added, and the datasets selection to which extra datasets are added. (so wind and rain would be included in your point ii - measurements which have no extra datasets)

File: MEDMI_Data_Download/met_extraction_script.py (hunk 3 above)
Comment (AnnAnnFryingPan) [reply MEDMI_Data_Download/met_extraction_script.py]: As were still finalising extra data, I've put some text in README, with a placeholder 'EDIT REQUEST' for once we know.

File: MEDMI_Data_Download/met_extraction_script.py [hunk 8]
```diff
@@ -67,84 +61,79 @@ def set_region(self, latitude_range, longitude_range):
...
             print('extracting data for {}'.format(self.MEASUREMENT_NAME))
         elif self._verbose > 1:
-            print('extracting data for {} using dict: {}\n and settings: {}'.format(
-                self.MEASUREMENT_NAME, json.dumps(dict, default=str), json.dumps(settings)))
-        datadata = self._perform_extraction(dict)
+            print('using settings: {}'.format( json.dumps(settings)))
+        datadata = self._perform_extraction(extraction_dict)
+        if save_to_file:
+            self._save_to_file(datadata, settings)
+        return datadata
```
Comment (douglowe) [inline MEDMI_Data_Download/met_extraction_script.py:200]: With the first logic check being >=, the second test of > is never going to trigger. Should the second test be made a separate if statement?

**Review by douglowe — COMMENTED** [review]
These look reasonable to me.

We should add an issue, for ensuring that the README's are all complete once we've finished all our changes to the code.

File: MEDMI_Data_Download/met_extraction_script.py (hunk 8 above)
Comment (AnnAnnFryingPan) [reply MEDMI_Data_Download/met_extraction_script.py:200]: Yes, fixed.
